File: src/main/java/no/nav/sosialhjelp/soknad/business/db/repositories/opplastetvedlegg/OpplastetVedleggRepositoryJdbc.java
//package no.nav.sosialhjelp.soknad.business.db.repositories.opplastetvedlegg;
//
//import no.nav.sosialhjelp.soknad.domain.OpplastetVedlegg;
//import no.nav.sosialhjelp.soknad.domain.VedleggType;
//import org.springframework.jdbc.core.RowMapper;
//import org.springframework.jdbc.core.namedparam.NamedParameterJdbcDaoSupport;
//import org.springframework.stereotype.Component;
//
//import javax.inject.Inject;
//import javax.inject.Named;
//import javax.sql.DataSource;
//import java.sql.ResultSet;
//import java.sql.SQLException;
//import java.util.List;
//import java.util.Optional;
//
//@Named("OpplastetVedleggRepository")
//@Component
//public class OpplastetVedleggRepositoryJdbc extends NamedParameterJdbcDaoSupport implements OpplastetVedleggRepository {
//
//    @Inject
//    public void setDS(DataSource ds) {
//        super.setDataSource(ds);
//    }
//
//    @Override
//    public Optional<OpplastetVedlegg> hentVedlegg(String uuid, String eier) {
//        return getJdbcTemplate().query("select * from OPPLASTET_VEDLEGG where EIER = ? and UUID = ?",
//                new OpplastetVedleggRowMapper(), eier, uuid).stream().findFirst();
//    }
//
//    @Override
//    public List<OpplastetVedlegg> hentVedleggForSoknad(Long soknadId, String eier) {
//        return getJdbcTemplate().query("select * from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?",
//                new OpplastetVedleggRowMapper(), eier, soknadId);
//    }
//
//    @Override
//    public String opprettVedlegg(OpplastetVedlegg opplastetVedlegg, String eier) {
//        if (eier == null || !eier.equalsIgnoreCase(opplastetVedlegg.getEier())) {
//            throw new RuntimeException("Eier stemmer ikke med vedleggets eier");
//        }
//        getJdbcTemplate()
//                .update("insert into OPPLASTET_VEDLEGG (UUID, EIER, TYPE, DATA, SOKNAD_UNDER_ARBEID_ID, FILNAVN, SHA512)" +
//                                " values (?,?,?,?,?,?,?)",
//                        opplastetVedlegg.getUuid(),
//                        opplastetVedlegg.getEier(),
//                        opplastetVedlegg.getVedleggType().getSammensattType(),
//                        opplastetVedlegg.getData(),
//                        opplastetVedlegg.getSoknadId(),
//                        opplastetVedlegg.getFilnavn(),
//                        opplastetVedlegg.getSha512());
//        return opplastetVedlegg.getUuid();
//    }
//
//    @Override
//    public void slettVedlegg(String uuid, String eier) {
//        getJdbcTemplate()
//                .update("delete from OPPLASTET_VEDLEGG where EIER = ? and UUID = ?",
//                        eier,
//                        uuid);
//    }
//
//    @Override
//    public void slettAlleVedleggForSoknad(Long soknadId, String eier) {
//        getJdbcTemplate()
//                .update("delete from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?",
//                        eier,
//                        soknadId);
//    }
//
//    @Override
//    public Integer hentSamletVedleggStorrelse(Long soknadId, String eier) {
//        if (getJdbcTemplate()
//                .queryForObject("select count(*) from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?", Integer.class, eier, soknadId) > 0) {
//            String sql = "select sum(dbms_lob.getLength(DATA)) from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?";
//            Integer totalSize = getJdbcTemplate()
//                    .queryForObject(sql, Integer.class, eier, soknadId);
//            return totalSize;
//        }
//        return 0;
//    }
//
//    public class OpplastetVedleggRowMapper implements RowMapper<OpplastetVedlegg> {
//
//        public OpplastetVedlegg mapRow(ResultSet rs, int rowNum) throws SQLException {
//            return new OpplastetVedlegg()
//                    .withUuid(rs.getString("uuid"))
//                    .withEier(rs.getString("eier"))
//                    .withVedleggType(new VedleggType(rs.getString("type")))
//                    .withData(rs.getBytes("data"))
//                    .withSoknadId(rs.getLong("soknad_under_arbeid_id"))
//                    .withFilnavn(rs.getString("filnavn"))
//                    .withSha512(rs.getString("sha512"));
//        }
//    }
//}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/db/repositories/opplastetvedlegg/OpplastetVedleggRepositoryJdbc.kt
package no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg

import no.nav.sosialhjelp.soknad.db.SQLUtils
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedleggRowMapper.opplastetVedleggRowMapper
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Repository

@Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
@Repository
open class OpplastetVedleggRepositoryJdbc(
    private val jdbcTemplate: JdbcTemplate
) : OpplastetVedleggRepository {

    override fun hentVedlegg(uuid: String?, eier: String): OpplastetVedlegg? {
        return jdbcTemplate.query(
            "select * from OPPLASTET_VEDLEGG where EIER = ? and UUID = ?",
            opplastetVedleggRowMapper,
            eier,
            uuid
        ).firstOrNull()
    }

    override fun hentVedleggForSoknad(soknadId: Long, eier: String?): List<OpplastetVedlegg> {
        return jdbcTemplate.query(
            "select * from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?",
            opplastetVedleggRowMapper,
            eier,
            soknadId
        )
    }

    override fun opprettVedlegg(opplastetVedlegg: OpplastetVedlegg, eier: String): String {
        if (!eier.equals(opplastetVedlegg.eier, ignoreCase = true)) {
            throw RuntimeException("Eier stemmer ikke med vedleggets eier")
        }
        jdbcTemplate.update(
            "insert into OPPLASTET_VEDLEGG (UUID, EIER, TYPE, DATA, SOKNAD_UNDER_ARBEID_ID, FILNAVN, SHA512) values (?,?,?,?,?,?,?)",
            opplastetVedlegg.uuid,
            opplastetVedlegg.eier,
            opplastetVedlegg.vedleggType.sammensattType,
            opplastetVedlegg.data,
            opplastetVedlegg.soknadId,
            opplastetVedlegg.filnavn,
            opplastetVedlegg.sha512
        )
        return opplastetVedlegg.uuid
    }

    override fun slettVedlegg(uuid: String?, eier: String) {
        jdbcTemplate.update("delete from OPPLASTET_VEDLEGG where EIER = ? and UUID = ?", eier, uuid)
    }

    override fun slettAlleVedleggForSoknad(soknadId: Long, eier: String) {
        jdbcTemplate.update(
            "delete from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?",
            eier,
            soknadId
        )
    }

    override fun hentSamletVedleggStorrelse(soknadId: Long, eier: String): Int {
        if (jdbcTemplate.queryForObject(
                "select count(*) from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?",
                Int::class.java, eier, soknadId
            ) > 0
        ) {
            val blobSize = SQLUtils.blobSizeQuery()
            return jdbcTemplate.queryForObject(
                "select sum($blobSize) from OPPLASTET_VEDLEGG where EIER = ? and SOKNAD_UNDER_ARBEID_ID = ?",
                Int::class.java,
                eier,
                soknadId
            )
        }
        return 0
    }
}
